File: yw/dom/ShadowRoot.lua
--[[
    Copyright (c) 2024, Oh Inseo (YJK) <dhdlstjtr@gmail.com>
    SPDX-License-Identifier: BSD-3-Clause
    This software may contain third-party material. For more info, see README.
]]
local object           = require "yw.common.object"
local DocumentFragment = require "yw.dom.DocumentFragment"


---https://dom.spec.whatwg.org/#concept-shadow-root
---@class DOM_ShadowRoot : DOM_DocumentFragment
---@field declarative                 boolean  https://dom.spec.whatwg.org/#shadowroot-declarative
---@field availableToElementInternals boolean  https://dom.spec.whatwg.org/#shadowroot-available-to-element-internals
local ShadowRoot = object.create(DocumentFragment)

---@param document DOM_Document
---@return DOM_ShadowRoot
function ShadowRoot:new(document)
    local o = DocumentFragment.new(self, document) --[[@as DOM_ShadowRoot]]
    o.isShadowRoot = true
    -- Shadow roots have an associated declarative (a boolean). It is initially set to false.
    o.declarative = false
    -- Shadow roots have an associated available to element internals. It is initially set to false.
    o.availableToElementInternals = false
    return o
end

return ShadowRoot

File: dom/shadow_root.go
// This file is part of YW project. Copyright 2025 Oh Inseo (YJK)
// SPDX-License-Identifier: BSD-3-Clause
// See LICENSE for details, and LICENSE.thirdparty.md for third-party license information.

package dom

// ShadowRoot represents a [DOM shadow root]
//
// [DOM shadow root]: https://dom.spec.whatwg.org/#concept-shadow-root
type ShadowRoot interface {
	DocumentFragment

	// CustomElementRegistry returns [custom element registry] of the shadow root.
	//
	// [custom element registry]: https://dom.spec.whatwg.org/#shadowroot-custom-element-registry
	CustomElementRegistry() *CustomElementRegistry

	// SetCustomElementRegistry sets [custom element registry] of the shadow root.
	//
	// [custom element registry]: https://dom.spec.whatwg.org/#shadowroot-custom-element-registry
	SetCustomElementRegistry(registry *CustomElementRegistry)
}
